// @glimmer/syntax/dist/modules/es2017/lib/generation/util.js
const ATTR_VALUE_REGEX_TEST = /[\xA0"&]/;
const ATTR_VALUE_REGEX_REPLACE = new RegExp(ATTR_VALUE_REGEX_TEST.source, 'g');
const TEXT_REGEX_TEST = /[\xA0&<>]/;
const TEXT_REGEX_REPLACE = new RegExp(TEXT_REGEX_TEST.source, 'g');

function attrValueReplacer(char) {
  switch (char.charCodeAt(0)) {
    case 160
    /* NBSP */
    :
      return '&nbsp;';

    case 34
    /* QUOT */
    :
      return '&quot;';

    case 38
    /* AMP */
    :
      return '&amp;';

    default:
      return char;
  }
}

function textReplacer(char) {
  switch (char.charCodeAt(0)) {
    case 160
    /* NBSP */
    :
      return '&nbsp;';

    case 38
    /* AMP */
    :
      return '&amp;';

    case 60
    /* LT */
    :
      return '&lt;';

    case 62
    /* GT */
    :
      return '&gt;';

    default:
      return char;
  }
}

export function escapeAttrValue(attrValue) {
  if (ATTR_VALUE_REGEX_TEST.test(attrValue)) {
    return attrValue.replace(ATTR_VALUE_REGEX_REPLACE, attrValueReplacer);
  }

  return attrValue;
}
export function escapeText(text) {
  if (TEXT_REGEX_TEST.test(text)) {
    return text.replace(TEXT_REGEX_REPLACE, textReplacer);
  }

  return text;
}
export function isSynthetic(node) {
  if (node && node.loc) {
    return node.loc.source === '(synthetic)';
  }

  return false;
}
export function sortByLoc(a, b) {
  // be conservative about the location where a new node is inserted
  if (isSynthetic(a) || isSynthetic(b)) {
    return 0;
  }

  if (a.loc.start.line < b.loc.start.line) {
    return -1;
  }

  if (a.loc.start.line === b.loc.start.line && a.loc.start.column < b.loc.start.column) {
    return -1;
  }

  if (a.loc.start.line === b.loc.start.line && a.loc.start.column === b.loc.start.column) {
    return 0;
  }

  return 1;
}
