// @checkup/core/lib/base-task.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const debug = require("debug");
const plugin_name_1 = require("./utils/plugin-name");
const config_1 = require("./config");
class BaseTask {
    constructor(pluginName, context) {
        this._pluginName = plugin_name_1.getShorthandName(pluginName);
        this.context = context;
        this.debug = debug('checkup:task');
    }
    get config() {
        this._parseConfig();
        return this._config;
    }
    get enabled() {
        var _a;
        this._parseConfig();
        let enabledViaFlag = ((_a = this.context.cliFlags.task) === null || _a === void 0 ? void 0 : _a.includes(this.fullyQualifiedTaskName)) || false;
        let enabled = this._enabledViaConfig || enabledViaFlag;
        this.debug('%s enabled: %s', this.fullyQualifiedTaskName, enabled);
        return enabled;
    }
    get fullyQualifiedTaskName() {
        return `${this._pluginName}/${this.taskName}`;
    }
    _parseConfig() {
        if (this._config) {
            return;
        }
        let config = this.context.config.tasks[this.fullyQualifiedTaskName];
        let [enabled, taskConfig] = config_1.parseConfigTuple(config);
        this._enabledViaConfig = enabled;
        this._config = taskConfig;
        this.debug('%s task config: %O', this.fullyQualifiedTaskName, this._config);
    }
}
exports.default = BaseTask;
