// @checkup/cli/lib/task-result-comparator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.taskResultComparator = void 0;
const DEFAULT_CATEGORIES = {
    metrics: 6,
    'best practices': 5,
    dependencies: 4,
    linting: 3,
    testing: 2,
    migrations: 1,
};
function getCategorySort(category) {
    var _a;
    return (_a = DEFAULT_CATEGORIES[category]) !== null && _a !== void 0 ? _a : -1;
}
function taskResultComparator(first, second) {
    var _a, _b, _c, _d;
    let firstC = (_a = first.properties) === null || _a === void 0 ? void 0 : _a.category;
    let firstGroup = ((_b = first.properties) === null || _b === void 0 ? void 0 : _b.group) || '';
    let secondC = (_c = second.properties) === null || _c === void 0 ? void 0 : _c.category;
    let secondGroup = ((_d = second.properties) === null || _d === void 0 ? void 0 : _d.group) || '';
    let firstCategory = getCategorySort(firstC);
    let secondCategory = getCategorySort(secondC);
    if (firstCategory === secondCategory) {
        // eslint-disable-next-line unicorn/no-nested-ternary
        return firstGroup > secondGroup ? -1 : firstGroup < secondGroup ? 1 : 0;
    }
    else {
        return firstCategory > secondCategory ? -1 : 1;
    }
}
exports.taskResultComparator = taskResultComparator;
